### PR TITLE
persist: introduce a new dynamic configuration library

### PR DESCRIPTION
**Pull request by danhhz**
Basic usage:
- A type-safe static `Config` is defined near where it is used.
- Once in the lifetime of a process, all interesting `Config`s are
  registered to a `ConfigSet`. The values within a `ConfigSet` are
  shared, though multiple `ConfigSet`s may be created and each are
  completely independent (i.e. one in each unit test).
- A `ConfigSet` is plumbed around as necessary and may be used to get
  or set the value of `Config`.

```rust
const FOO: Config<bool> = Config::new("foo", false, "description of foo");
fn bar(cfg: &ConfigSet) {
    assert!(FOO.get(&cfg), false);
}
fn main() {
    let cfg = ConfigSet::default.add(&FOO);
    bar(&cfg);
}
```

Design considerations for this library

- The primary motivation is minimal boilerplate. Runtime dynamic
  configuration is one of the most powerful tools we have to quickly
  react to incidents, etc in production. Adding and using them should be
  easy enough that engineers feel empowered to use them generously.

  The theoretical minimum boilerplate is 1) declare a config and 2) use a
  config to get/set the value. These could be combined into one step if
  (2) were based on global state, but that doesn't play well with
  testing. So instead we accomplish (2) by constructing a shared bag of
  config values in each `fn main`, amortizing the cost by plumbing it once
  to each component (not once per config).
- Config definitions are kept next to the usage. The common case is that
  a config is used in only one place and this makes it easy to see the
  associated documentation at the usage site. Configs that are used in
  multiple places may be defined in some common place as appropriate.
- Everything is type-safe.
- Secondarily: set up the ability to get and use the latest values of
  configs in tooling like `persistcli` and `stash debug`. As we've
  embraced dynamic configuration, we've ended up in a situation where
  it's stressful to run the read-write `persistcli admin` tooling with
  the defaults compiled into code, but `persistcli` doesn't have access
  to the vars stuff and doesn't want to instantiate a catalog impl.

This design is very similar to a generalized version of the `PersistFeatureFlag` that Ben added. Interestingly, because this started as a very old branch, I didn't use it as a reference until the similarity was pointed out to me. This is a fail on my part, but the upside is that it's additional confirmation of the design!

Additionally, #24349 is an alternative to this PR and is a more gradual evolution.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Tips for reviewer

Accepting naming bikesheds!

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
Added @benesch as a reviewer because I think it's worth pulling this out into mz_ore or mz_dyncfg and using it everywhere (as a followup).

File: src/persist-client/src/cfg.rs
Comment (danhhz) [inline src/persist-client/src/cfg.rs:272]: This is a third touchpoint when adding a flag in addition to the two I mention in the "theoretical minimum" section.

After letting this mull, I'm a pretty strong yes on replacing it with a static global registry at this point. Basically, the boilerplate has gotten light enough that any additional work now feels meaningful in comparison. However not doing that in this PR because (1) IIUC Ben is a skeptic of the global state and (2) I'm not sure I understand how sketch the crates available to do that are.

For (1), while I strongly agree that flag _values_ should not be global state, I have a harder time making an argument that global state containing the set of flag _definitions_ in a binary is something to worry about in practice. However, the one worry I do have here is that a config defined in code that exists only in clusterd (e.g. much of compute) wouldn't get picked up by envd and the explicit declarations currently here would prevent that subtlety.

For (2), we do have one of the possibly-sketchy crates [already used in mz_ore](https://github.com/MaterializeInc/materialize/blob/d11c694ef8d207c74ad8afa1aa0a026de02f0a20/src/ore/src/assert.rs#L60). I proved to myself in a local branch that we could use this one already in the dep graph for configs as long as it's possible to use a Mutex in the `ctor` section. There's also something like the `inventory` crate which is essentially designed exactly for this registration sort of use.

Comment (bkirwi) [reply src/persist-client/src/cfg.rs:272]: > IIUC Ben is a skeptic of the global state

Yeah, still true! Basically I find the idea of code's behaviour changing depending on what's happens to be linked in to be sort of alarming and hard to reason about. Like, suppose I hit the `clusterd` bug you're talking about by taking a function with only one caller and moving it next to the code that calls it; normally a harmless refactoring, but now I've just accidentally changed materialize's SQL API. Spooky!

Conversely, I think the one extra line just bothers me less. (Which I appreciate is a matter of taste!)

**Comment by bkirwi** [comment]
Linking up my PR from earlier this week: https://github.com/MaterializeInc/materialize/pull/24349

I think they're more similar than different; and I'd definitely prefer either of them to the status quo.

This PR has a larger diff and a wider radius. I think that's mostly due to this PR trying to be a candidate for `mz_ore`, while the other PR is a bit narrower? I agree that it'd be nice to pull this into `mz_ore`... aside from having less Persist stuff, I think it's a bit silly that every team ends up with its own duplicative `MyParameters` struct, and we're probably all better off if defining vars is extremely easy for everyone.

File: src/persist-client/src/cfg.rs
Comment (benesch) [reply src/persist-client/src/cfg.rs:272]: > However, the one worry I do have here is that a config defined in code that exists only in clusterd (e.g. much of compute) wouldn't get picked up by envd and the explicit declarations currently here would prevent that subtlety.

Hm, I was also in favor of the global flag registry until you mentioned this footgun. That seems somewhat hard to work around? 

One idea that comes to mind: when `envd` initializes its connection with `clusterd`, it sends over all the list of all known flags. If `clusterd` knows of any flags that `envd` doesn't send over, then `clusterd` panics with an informative message. That seems possibly workable!

File: src/persist-client/src/batch.rs
Comment (benesch) [inline src/persist-client/src/batch.rs:232]: Style thought: the `new` constructor feels a little Go-y—looks a lot like the Go flags package. I feel like it'd be slightly more Rust-y to use a params struct:

```rust
Config::new(ConfigSpec {
    name: "persist_batch_delete_enabled",
    default: false,
    description: "Whether to actually delete blobs when batch delete is called (Materialize).",
});
```

It's slightly more work to type out and slightly more verbose, but I find it slightly faster to understand what each of the parameters is while reading the code.

**Review by benesch — COMMENTED** [review]
> Added @benesch as a reviewer because I think it's worth pulling this out into mz_ore or mz_dyncfg and using it everywhere (as a followup).

I'm not going to have time to review in depth, but I'm very, very supportive of this effort, and the goal of making this the blessed way to add feature flags across the entire codebase. 🙌🏽 

I'd vote for putting this into a new crate like `mz_dyncfg` rather than `mz_ore`. (Also we should decide whether it's `dyncfg` or `dyn_cfg` and use whichever consistently. I have no opinion on which spelling is better.)

**Review by mjibson — APPROVED** [review]
LGTM, assuming we can at some point move the generic stuff out of persist-client. Seems fine for now to have it all live there for the initial PR.

File: src/sql/src/session/vars.rs
Comment (bkirwi) [inline src/sql/src/session/vars.rs:3811]: FWIW I think this is basically fine, at least for now!

**Comment by bkirwi** [comment]
A couple comments, but I think the only non-nit request is to add a `const WHATEVER: Config<String>` and shake out any issues that exposes...

**Review by def- — COMMENTED** [review]
One thing that came to mind: If we categorize the config options, it might be nice to have a `--future-default-flags` to enable all flags that we plan to enable soon, so that we get them covered in CI automatically, instead of the current approach of maintaining such a list in https://github.com/MaterializeInc/materialize/blob/151a5c7cb8d37d8fd4e2da482865fb5da03c2846/misc/python/materialize/mzcompose/__init__.py#L41 manually.

File: src/persist-client/src/dyn_cfg.rs
Comment (danhhz) [reply src/persist-client/src/dyn_cfg.rs:270]: Done!

File: src/persist-client/src/dyn_cfg.rs
Comment (danhhz) [reply src/persist-client/src/dyn_cfg.rs:315]: Done

File: src/persist-client/src/dyn_cfg.rs
Comment (danhhz) [reply src/persist-client/src/dyn_cfg.rs:128]: I did this because when you're getting a config value, the critical "noun" is more which config than the bag of values. Concretely, I find

```
if STREAMING_SNAPSHOT_AND_FETCH_ENABLED.get(&self.machine.applier.cfg.configs) {
```

easier to read than

```
if self.machine.applier.cfg.configs.get(&STREAMING_SNAPSHOT_AND_FETCH_ENABLED) {
```

Secondarily, I _think_ rustfmt might do a worse job on the latter (break it up into a bunch of lines and make it hard to follow the control flow).

That said, not a hill I'd die on! I'm curious if anyone else on this PR has a strong opinion one way or the other?

**Review by danhhz — COMMENTED** [review]
Realized that I forgot to post my initial plan for overall timeline:

- 1: Get consensus on new design and merge it (this PR).
- 2: Swap over all persist configs. Use this to shake out and work through things like the `&const String` Ben mentioned, `Option<T>` etc.
- 3: Pull this out of persist, so everyone can use it, and socialize. (Not to convert the whole codebase myself.)

TBD is if we have to resolve the global registry one way or the other before (3). I think no?

> I'd vote for putting this into a new crate like `mz_dyncfg` rather than `mz_ore`. (Also we should decide whether it's `dyncfg` or `dyn_cfg` and use whichever consistently. I have no opinion on which spelling is better.)

Cool, that's my vote, too. We can hash out the naming in the PR to extract this out of persist.

> Linking up my PR from earlier this week: #24349
> 
> I think they're more similar than different; and I'd definitely prefer either of them to the status quo.
> 
> This PR has a larger diff and a wider radius. I think that's mostly due to this PR trying to be a candidate for `mz_ore`, while the other PR is a bit narrower? I agree that it'd be nice to pull this into `mz_ore`... aside from having less Persist stuff, I think it's a bit silly that every team ends up with its own duplicative `MyParameters` struct, and we're probably all better off if defining vars is extremely easy for everyone.

Ah, I meant to link that as an alternative and also call out the similarity to PersistFeatureFlag in the PR description, but dropped the ball on this. Fixed now!

> One thing that came to mind: If we categorize the config options, it might be nice to have a `--future-default-flags` to enable all flags that we plan to enable soon, so that we get them covered in CI automatically, instead of the current approach of maintaining such a list in misc/python/materialize/mzcompose/__init__.py manually.

Added a TODO for the categorization and included this idea in it.

File: src/persist-client/src/batch.rs
Comment (benesch) [reply src/persist-client/src/batch.rs:232]: Weird! Seems like a `rustc` bug, or at least a limitation.

> Another option is a `config_{int,..}!` macro, which might anyway be something we need to add if we do the global static registry. Ditto if we want the file and line number of the config definition to be available.

Makes sense. Happy to defer this conversation to a later PR that adds the global registry.

File: src/persist-client/src/cfg.rs
Comment (benesch) [reply src/persist-client/src/cfg.rs:272]: > don't think we need to block on the specifics here

Works for me!

File: src/persist-client/src/dyn_cfg.rs
Comment (bkirwi) [reply src/persist-client/src/dyn_cfg.rs:254]: Poked at it.

If we wanted to wrap the `BTreeMap`, it looks like the critical annoyance is `ConfigSet::add`, which is the only place that we require exclusive access to the map. Since we do all our adds at the beginning, this isn't a dealbreaker necessarily -- the fancy way to solve this would be to have a `ConfigSetBuilder`, I guess? Up to you whether that's worth the hassle.

File: src/persist-client/src/dyn_cfg.rs
Comment (bkirwi) [reply src/persist-client/src/dyn_cfg.rs:365]: I do think it would be fine to punt on `String` support... but this does feel like a weird interim state, where it's sort-of supported, but anyone who tries to actually declare a `String` config will be unable to do so without reworking the core trait.

If you'd rather leave it for followup work, leaving a warning comment and/or dropping this impl also seem fine to me.

File: src/sql/src/session/vars.rs
Comment (danhhz) [reply src/sql/src/session/vars.rs]: Leaned into this and removed the HACK

File: src/persist-client/src/batch.rs
Comment (danhhz) [reply src/persist-client/src/batch.rs:232]: > Weird! Seems like a rustc bug, or at least a limitation.

Yeah, I got the same vibes.

> Happy to defer this conversation to a later PR that adds the global registry.

I might take you up on this. Another option (disclaimer: that I haven't tried, no idea if this works) is a sketchy `std::mem::forget` and a comment in `new`.

File: src/persist-client/src/dyn_cfg.rs
Comment (danhhz) [reply src/persist-client/src/dyn_cfg.rs:254]: Tried this. Ended up running into the `Config::shared` method, which I realized was my original answer to this issue! (The hazards of only working on something on Fridays, I suppose.) That allows perf-sensitive users to both avoid this as well and do the lookup of the name in the BTreeMap only once. It does double the number of lines of code needed for a get from 1 to 2, but my thinking was that was reasonable for uses that opt in to being more perf-sensitive.

Thoughts? Also happy to remove that method and go with the `ConfigSetBuilder` if we think we'd rather optimize the common-case a bit in exchange for losing the ability to amortize the name lookup.

Comment (bkirwi) [reply src/persist-client/src/dyn_cfg.rs:254]: Eh... I suspect it's rare that someone will take advantage of the two-step thing - hot enough to notice a map lookup, not hot enough to care about an atomic read, don't want to cache - but I don't mind punting and seeing how it gets picked up in practice!

**Review by bkirwi — APPROVED** [review]

File: src/persist-client/src/dyn_cfg.rs
Comment (bkirwi) [reply src/persist-client/src/dyn_cfg.rs:367]: Cool thanks! And yeah, definitely easy enough to tweak later if we need to...

**Pull request merged**
